src/darwin/frida-helper-backend-glue.m

```diff
@@ -1556,26 +1567,284 @@ static gboolean frida_agent_context_init_functions (FridaAgentContext * self, Gu
+      initialized_address = (*info_ptr) + 17;
+      g_free (info_ptr);
+
+      break;
+    }
+
+    case GUM_CPU_ARM64:
+    case GUM_CPU_AMD64:
+    {
+      guint64 * info_ptr;
+
+      info_ptr = (guint64 *) gum_darwin_read (self->task, self->info_address, sizeof (info_ptr), NULL);
+      initialized_address = (*info_ptr) + 25;
+      g_free (info_ptr);
+
+      break;
+    }
+
+    default:
+      g_assert_not_reached ();
+  }
+
+  bret = gum_darwin_write (self->task, initialized_address, &yes, 1);
+  g_assert_cmpint (bret, ==, TRUE);
+}
+
+static gboolean
+frida_spawn_instance_is_libc_initialized (FridaSpawnInstance * self)
+{
+  gboolean initialized;
+  GumAddress initialized_address;
+  guint8 * yes;
+
+  switch (self->cpu_type)
+  {
+    case GUM_CPU_ARM:
+    case GUM_CPU_IA32:
+    {
+      guint32 * info_ptr;
+
+      info_ptr = (guint32 *) gum_darwin_read (self->task, self->info_address, sizeof (info_ptr), NULL);
+      initialized_address = (*info_ptr) + 17;
+      g_free (info_ptr);
+
+      break;
+    }
+
+    case GUM_CPU_ARM64:
+    case GUM_CPU_AMD64:
+    {
+      guint64 * info_ptr;
+
+      info_ptr = (guint64 *) gum_darwin_read (self->task, self->info_address, sizeof (info_ptr), NULL);
+      initialized_address = (*info_ptr) + 25;
+      g_free (info_ptr);
+
+      break;
+    }
+
+    default:
+      g_assert_not_reached ();
+  }
+
+  yes = (guint8 *) gum_darwin_read (self->task, initialized_address, sizeof (yes), NULL);
+  initialized = *yes;
+  g_free (yes);
+
+  return initialized;
+}
+
+static void
 frida_spawn_instance_on_server_recv (void * context)
 {
   FridaSpawnInstance * self = context;
+  kern_return_t ret;
+  thread_state_flavor_t state_flavor = GUM_DARWIN_THREAD_STATE_FLAVOR;
+  mach_msg_type_number_t state_count = GUM_DARWIN_THREAD_STATE_COUNT;
+  GumDarwinUnifiedThreadState state;
 
   frida_spawn_instance_receive_breakpoint_request (self);
 
   switch (self->breakpoint_phase)
   {
-    case FRIDA_BREAKPOINT_FIRST:
-      frida_set_debug_state (self->thread, &self->second_debug_state, self->cpu_type);
-      self->breakpoint_phase = FRIDA_BREAKPOINT_SECOND;
+    case FRIDA_BREAKPOINT_SET_HELPERS:
+      ret = thread_get_state (self->thread, state_flavor, (thread_state_t) &state, &state_count);
+      g_assert_cmpint (ret, ==, KERN_SUCCESS);
+
+      memcpy (&self->previous_thread_state, &state, sizeof (state));
+
+      frida_spawn_instance_call_set_helpers (self, state, self->fake_helpers);
+
+      self->breakpoint_phase = FRIDA_BREAKPOINT_DLOPEN_LIBC;
+      frida_spawn_instance_send_breakpoint_response (self);
+
+      break;
+
+    case FRIDA_BREAKPOINT_DLOPEN_LIBC:
+      if (frida_spawn_instance_is_libc_initialized (self))
+        frida_spawn_instance_unset_helpers (self);
+
+      memcpy (&state, &self->previous_thread_state, sizeof (state));
+
+      frida_spawn_instance_call_dlopen (self, state, self->lib_name, RTLD_LAZY | RTLD_GLOBAL);
```

Comment:
Nit: we usually specify these two flags in this order: `RTLD_GLOBAL | RTLD_LAZY`.